api/src/main/java/org/openmrs/module/xdsbrepository/impl/XDSbServiceImpl.java

```diff
@@ -179,6 +181,41 @@ public RegistryResponseType provideAndRegisterDocumentSetB(ProvideAndRegisterDoc
 		return response;
 	}
 
+	protected void validateDocumentMatchMetadata(List<ExtrinsicObjectType> extrinsicObjects, Map<String, ProvideAndRegisterDocumentSetRequestType.Document> documents) throws XDSException {
+		Set<String> metadataIds = new HashSet<String>();
+		for (ExtrinsicObjectType eot : extrinsicObjects) {
+			metadataIds.add(eot.getId());
+		}
+
+		Set<String> allIds = new HashSet<String>();
+		allIds.addAll(metadataIds);
+		allIds.addAll(documents.keySet());
+
+		Set<String> missingDocs = new HashSet<String>(allIds);
+		missingDocs.removeAll(documents.keySet());
+		Set<String> missingMetadata = new HashSet<String>(allIds);
+		missingMetadata.removeAll(metadataIds);
+
+		if (missingDocs.size() > 0) {
+			throw new XDSException(XDSException.XDS_ERR_MISSING_DOCUMENT, "The following documents are referenced by metadata but are missing: " + stringifySet(missingDocs), null);
+		}
+		if (missingMetadata.size() > 0) {
+			throw new XDSException(XDSException.XDS_ERR_MISSING_DOCUMENT_METADATA, "The following documents were found but their metadata is missing: " + stringifySet(missingMetadata), null);
+		}
+	}
+
+	private String stringifySet(Set<String> missingDocs) {
```

Comment:
You might be able to use the `join` method from apache `StringUtils`, cut it down to a one-liner - it should already be in included in the project scope.